Embed pause image

bpradipt opened the pull request.

bpradipt committed:
podvm-builder: Install skopeo and umoci

Both these packages are needed for downloading and embedding
container images in the podvm image

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
podvm-builder: Add support for embedding pause image

Downloads the pause image using skopeo and uses
umoci to create the oci bundle for embedding

The following OCP pause image is embedded by default
quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:7f3cb6f9d265291b47a7491c2ba4f4dd0752a18b661eee40584f9a5dbcbe13bb

A new bash method download_and_extract_pause_image is introduced which
accepts three parameters via env variable:

PAUSE_IMAGE_REPO, PAUSE_IMAGE_VERSION and PAUSE_IMAGE_REPO_AUTH_FILE

These can be set in the respective podvm image configMap
(<provider>-podvm-image-cm.yaml) before kickstarting the pod vm image
generation if required.

The PAUSE_IMAGE_REPO_AUTH_FILE points to a json file containing the registry
secret to pull the image from authenticated registry.

The default is the cluster secret that is mounted as a volume in the
`/tmp/regauth/auth.json` for osc-podvm-create-job.yaml to download OCP
pause image.

Fixes: #KATA-3156

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
podvm-builder: Update the base image to ubi:9.4

Latest stable image with security fixes

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
podvm-builder: Update labels to align with upstream

cloud-api-adaptor uses kata/main now. So update the labels
to reflect the current versions of kata used with CAA

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
podvm-builder: Adapt the src layout based on upstream changes

The upstream code layout changed from src/cloud-api-adaptor
to src/cloud-api-adaptor/src/cloud-api-adaptor. Accordingly
change the handling of source code directory.

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt commented:
@savitaashture fyi

snir911 commented on config/peerpods/podvm/aws-podvm-image-cm.yaml:
```diff
@@ -33,3 +33,8 @@ data:
+
+  # Pause image vars
+  PAUSE_IMAGE_REPO: "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256"
+  PAUSE_IMAGE_VERSION: "ac97982e845413b2b7377171d1b4c85d2b2cfda174e8251e3049d4843cea40e1"
```
would this be different for other architectures?

bpradipt replied:
I think so

bpradipt replied:
The OCP pause image is multi-arch.

snir911 commented on config/peerpods/podvm/aws-podvm-image-cm.yaml:
```diff
@@ -33,3 +33,8 @@ data:
+  # Pause image vars
+  PAUSE_IMAGE_REPO: "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256"
+  PAUSE_IMAGE_VERSION: "ac97982e845413b2b7377171d1b4c85d2b2cfda174e8251e3049d4843cea40e1"
+  PAUSE_IMAGE_REPO_AUTH_FILE: "/tmp/regauth/auth.json"
```
Wouldn't it be preferred to hide these from the user, i.e. set default values withing the scripts and comment these variables out?
Alternatively, can we simply use by default the current pause image used normally by Openshift?

bpradipt replied:
I agree.. If there is way to query the default pause for a release then we can set it in the operator, like we do for some other vars. Otherwise we can set it to a default (like above).

bpradipt replied:
@snir911 I couldn't find a way to query this info. One option is to mount the `/etc/crio/crio.conf.d/00-default` and read it. But it'll increase the required permissions for the container so I prefer to avoid it. 
For now, my suggestion is to use this approach of setting the default in configmap. Since this is a generic pause image it should work with other OCP versions as well. Having these options in the configmap will make it easier to replace it - using smaller size pause image, different one for a specific OCP release etc

snir911 replied:
you can fetch it from the worker node mc, however, this indeed may be an overkill, i think it would be better to set it to some fixed value in the scripts (if not explicitly set) rather than having it in the cm. 
This will reduce confusion of users when customize cm image is required.

snir911 replied:
oh, downside will be it will require to explicitly set for different architectures...

bpradipt replied:
So as you suggested, I'll set the default per arch (based on current info) in the script for now and leave it there..
Wdyt ?

bpradipt replied:
Btw I didn't find the pause image details in the worker mc :-(

snir911 commented on config/peerpods/podvm/lib.sh:
```diff
@@ -61,13 +62,15 @@ function install_rpm_packages() {
         "packer=https://releases.hashicorp.com/packer/1.9.4/packer_1.9.4_linux_amd64.zip"
         "kubectl=https://mirror.openshift.com/pub/openshift-v4/x86_64/clients/ocp/4.14.9/openshift-client-linux.tar.gz"
         "yq=https://github.com/mikefarah/yq/releases/download/v4.35.2/yq_linux_amd64.tar.gz"
+        "umoci=https://github.com/opencontainers/umoci/releases/download/v0.4.7/umoci.amd64"
```
for later, maybe worth to validate hashs at some point

bpradipt replied:
agree

bpradipt committed:
podvm-image: Handle RELATED_IMAGE_PODVM_PAYLOAD in the code

When RELATED_IMAGE_PODVM_PAYLOAD is set, update the job spec to use
this image in initContainer.

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

snir911 commented on config/peerpods/podvm/lib.sh:
```diff
@@ -61,13 +62,15 @@ function install_rpm_packages() {
 #"packer=https://releases.hashicorp.com/packer/1.9.4/packer_1.9.4_linux_amd64.zip"
 #"kubectl=https://mirror.openshift.com/pub/openshift-v4/x86_64/clients/ocp/4.14.9/openshift-client-linux.tar.gz"
 #"yq=https://github.com/mikefarah/yq/releases/download/v4.35.2/yq_linux_amd64.tar.gz"
+#"umoci=https://github.com/opencontainers/umoci/releases/download/v0.4.7/umoci.amd64"
```
we'll need a way to configure it once multiarch is supported

bpradipt replied:
Yeah, all the binaries are x86_64 for now and needs to be switched to multi-arch. I'll add a TBD

snir911 commented on config/peerpods/podvm/lib.sh:
```diff
@@ -187,6 +187,55 @@ function prepare_source_code() {
+
+    # Set default values for the OCP pause image
+    # This image is multi-arch
+    PAUSE_IMAGE_REPO_DEFAULT="quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256"
```
nit: maybe define the defaults in the top of the file?

bpradipt replied:
yeah, I thought about it but then not sure what made me put here ;) .. I think I was in two minds whether to put it at the top or in the function..

snir911 commented on controllers/image_generator.go:
```diff
@@ -308,6 +308,15 @@ func (r *ImageGenerator) createJobFromFile(jobFileName string) (*batchv1.Job, er
+	// If RELATED_PODVM_PAYLOAD_IMAGE environment variable is set, use it
+	// Otherwise, use the default podvm payload image
+	// There is only one initContainer in the job, so we don't need to check the container name
+	podvmPayloadImage := os.Getenv("RELATED_IMAGE_PODVM_PAYLOAD")
```
Would it work if it's being set in the podvm-image-cm? IIUC it's not, and i think it may be a bit confusing that this is image related configuration only but it's set differently

bpradipt replied:
The RELATED_* is also used for disconnected installs (useful for mirroring). Hence we use the same thing.

snir911 reviewed (approved):
LGTM, thanks! added some minor comments

bpradipt commented:
Tested with custom operator and podvm-builder image to generate a podvm image with embedded pause image in Azure
